project/LienShark/Readme.rst

```diff
@@ -0,0 +1,274 @@
+
+
+######################
+Problem to Solve
+######################
+
+First, I will learn to perform SVD and spectral decomposition for PCA.  
+I will also survey current methods, comparing and improving them.  
+In the second phase, I hope to integrate the Gersgorin Theorem to compare results,  
+aiming to trade some accuracy for increased speed.
+
+This project has two main objectives:
+1. **Speeding up PCA and SVD decomposition.**
+2. **Addressing the challenge of eigenvalue computation and dimensionality reduction for large matrices.**
+
+These challenges are common in machine learning, numerical linear algebra, and scientific computing,  
+especially in areas such as NLP, genetic data analysis, and big data processing. As dataset sizes increase,  
+existing tools often struggle with speed and memory limitations.
+
+By applying the Gersgorin Theorem, I will estimate eigenvalues for large matrices and use PCA for dimensionality reduction.
+
+### Introduction to the Gersgorin Theorem:
+The Gersgorin Theorem approximates the eigenvalues of matrices.  
+It states that all eigenvalues of a matrix lie within specific disks (called Gersgorin disks) on the complex plane.
+
+
+######################
+Prospective Users
+######################
+
+I expect this approach to shorten computation times and improve efficiency in data-intensive research,  
+such as genetic data analysis and text analysis of large corpora.  
+Since my approach sacrifices some accuracy for faster speed, it could be especially valuable for tasks like large-scale NLP training on personal computers.
+
+Additionally, I will explore applications in financial modeling, machine learning, and big data processing.  
+My goal is to reduce memory and computing costs during data processing, giving the project practical application value.
+
+The baseline methods will include traditional SVD and spectral decomposition methods, which I aim to improve.
+
+
+######################
+System Architecture
+######################
+
+**1. Input and Output:**
+
+- **Input:**  
+  Massive matrices.  
+  For larger matrices, I plan to experiment with either Incremental SVD or memory mapping to address memory limitations.  
+  I will start with smaller matrices to ensure feasibility but aim to work with million-row matrices eventually.  
+  If that proves impractical, I will adjust my approach accordingly.
+
+  I will adopt the row-major approach to store my matrix because, in most cases, a row in a matrix typically represents a new data sample,
+  while a column represents its features. When performing Incremental SVD, I tend to add more rows incrementally, as this approach is more meaningful in practice. 
+  It also helps reduce computational complexity and simplifies storage management.
+  For e.g: Today, if I need to process a matrix larger than 1000x1000, I would first load a portion of it, and then incrementally update the SVD one row at a time
```

Comment:
The progressive reading (updating rows/columns one by one) works for both row- and column-majoring.  Do not hold the wrong idea that row-majoring is better for the progressive reading.